<div class="alert alert-success">
<b> Комментарий от ревьюера 2 ✔️</b>

Привет еще раз. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<b>Комментарий от ревьюера</b>

Привет! Меня зовут Ярослав Кальманов, я буду делать ревью твоего проекта. Давай будем общаться на «ты».

Мои комментарии будут в ячейках "Комментарий от ревьюера".

<div class="alert alert-danger"> 

<b> Комментарий от ревьюера ❗</b>

Красным цветом обозначены комментарии, на которые тебе нужно обратить внимание в первую очередь, чтобы я принял проект. 
</div>
<div class="alert alert-warning">

<b> Комментарий от ревьюера ⚠️</b>

Желтым цветом отмечено то, что ты можешь сделать немного по-другому. Ты можешь учесть эти комментарии при выполнении следующих проектов.
</div>
<div class="alert alert-success">

<b> Комментарий от ревьюера ✔️</b>

Зелёным цветом обозначены твои удачные решения, которые ты без сомнения можешь применять в будущих проектах.  
</div>

Если ты вносишь изменения в проект по моим рекомендациям, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь свои комментарии заметным цветом или поместишь 

<div class="alert alert-info">
    
в такую ячейку
</div>

Давай при работе с проектом мы будем общаться с тобой в диалоге, поэтому не удаляй и не перемещай мои комментарии, это поможет мне сделать повторную  проверку проекта быстрее.
___

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Изучение общей информации о данных

**Описание данных:**  
children — количество детей в семье  
days_employed — общий трудовой стаж в днях  
dob_years — возраст клиента в годах   
education — уровень образования клиента  
education_id — идентификатор уровня образования  
family_status — семейное положение  
family_status_id — идентификатор семейного положения  
gender — пол клиента  
income_type — тип занятости  
debt — имел ли задолженность по возврату кредитов  
total_income — ежемесячный доход  
purpose — цель получения кредита  

In [1]:
import pandas as pd

data = pd.read_csv('C:\Users\Dao\YandexDisk\IT\Datasets\yandex_bank.csv')

# общая информация о датасете
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [2]:
# общие статистические данные
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [3]:
# первые 15 строк датасета

data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [4]:
# последние 15 строк датасета

data.tail(15)

children  days_employed  dob_years            education  education_id  \
21510         2            NaN         28              среднее             1   
21511         0    -612.569129         29               высшее             0   
21512         0    -165.377752         26               высшее             0   
21513         0   -1166.216789         35              среднее             1   
21514         0    -280.469996         27  неоконченное высшее             2   
21515         1    -467.685130         28              среднее             1   
21516         0    -914.391429         42               высшее             0   
21517         0    -404.679034         42               высшее             0   
21518         0  373995.710838         59              СРЕДНЕЕ             1   
21519         1   -2351.431934         37       ученая степень             4   
21520         1   -4529.316663         43              среднее             1   
21521         0  343937.404131         67              среднее             1   
21522         1   -2113.346888         38              среднее             1   
21523         3   -3112.481705         38              среднее             1   
21524         2   -1984.507589         40              среднее             1   

               family_status  family_status_id gender income_type  debt  \
21510        женат / замужем                 0      F   сотрудник     0   
21511       гражданский брак                 1      F   сотрудник     1   
21512  Не женат / не замужем                 4      M   компаньон     0   
21513        женат / замужем                 0      F   сотрудник     0   
21514  Не женат / не замужем                 4      M   компаньон     0   
21515        женат / замужем                 0      F   сотрудник     1   
21516        женат / замужем                 0      F   компаньон     0   
21517       гражданский брак                 1      F   компаньон     0   
21518        женат / замужем                 0      F   пенсионер     0   
21519              в разводе                 3      M   сотрудник     0   
21520       гражданский брак                 1      F   компаньон     0   
21521        женат / замужем                 0      F   пенсионер     0   
21522       гражданский брак                 1      M   сотрудник     1   
21523        женат / замужем                 0      M   сотрудник     1   
21524        женат / замужем                 0      F   сотрудник     0   

        total_income                                purpose  
21510            NaN                приобретение автомобиля  
21511  140068.472941                покупка жилья для сдачи  
21512  147301.457769  получение дополнительного образования  
21513  250986.142309                          покупка жилья  
21514  355988.407188             строительство недвижимости  
21515  109486.327999                  заняться образованием  
21516  322807.776603                   покупка своего жилья  
21517  178059.553491           на покупку своего автомобиля  
21518  153864.650328                   сделка с автомобилем  
21519  115949.039788      покупка коммерческой недвижимости  
21520  224791.862382                      операции с жильем  
21521  155999.806512                   сделка с автомобилем  
21522   89672.561153                           недвижимость  
21523  244093.050500           на покупку своего автомобиля  
21524   82047.418899                  на покупку автомобиля

**Выводы**

1. Данные описывают семейное положение, социально-экономический статус и кредитоспособность клиентов.
2. Имена столбцов без нарушений стиля.
3. В столбцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход) имеются пропуски. 
4. В столбце days_employed (общий трудовой стаж в днях) и children (количество детей в семье) присутствуют отрицательные значения. Это могло произойти из-за ошибки заполнения данных, когда значение записывается с тире перед ним.
5. В столбцах education (уровень образования клиента) и family_status (семейное положение) имеются строки в верхнем регистре.
6. Максимальное значение трудового стажа (столбец days_employed): 401755 дней - это 1100 лет. Из чего можно сделать предположение, что в ряде значений трудовой стаж записан не в днях, а в часах. 401755 часов - это приблизительно 16 739 дней или 46 лет. 
7. Поскольку минимальное значение в столбце dob_years (возраст клиента в годах) равняется 0, это значит, что часть значений были заполнены плейсхолдерами (нулями) вместо NaN.
8. В столбце children (количество детей в семье) максимальное значение детей маловероятно высокое: 20. Такая цифра, предположительно, могла возникнуть в результате технической ошибки в процессе переноса данных, когда изначально число, записанное как вещественное - 2.0 (два ребенка), не потеряло ноль в результате форматирования даннных. 

Данных достаточно для анализа надёжности заёмщиков, но необходима предварительная предобработка данных.


<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Все верно
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

Найдем и заменим все пропуски типа NaN.

In [5]:
# Найдем пропуски во всем датасете

data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
# определим присутствуют ли пропуски в сболбцах days_employed и total_income в одних и тех же строках

gaps = data.fillna(value=0)
gaps = gaps.loc[(gaps['days_employed'] == 0) & (gaps['total_income'] == 0)]
print('В данных', gaps.shape[0], 'строк с пропусками одновременно в обоих столбцах - days_employed и total_income')

В данных 2174 строк с пропусками одновременно в обоих столбцах - days_employed и total_income


In [7]:
# посмотрим какие типы занятости присутствуют в датафрейме

data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

In [8]:
# проверим соответствие пропусков типу занятости

print('Количество пропусков по типу занятости:')
data.loc[data['days_employed'].isna(), 'income_type'].value_counts()

Количество пропусков по типу занятости:


сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

Таким образом, пропуски не являются случайными, а представлены во всех основных типах занятости одновременно в двух столбцах: трудовой стаж и общий доход и составляют приблизительно 10% в каждом типе занятости. Из чего можно предположить, что клиенты банка намеренно скрыли эти данные. В качестве основной причины такого сокрытия можно предположить отсутствие официального источника дохода и, соответственно, трудового стажа. В случае госслужащих намеренное сокрытие могло быть связано с другими причинами, например, с государственной тайной (служащие спецслужб и пр.).

Часть значений в столбце days_employed являются отрицательными. Приведем их в положительное состояние.  

In [9]:
# для этого вернем абсолютные значения чисел с помощью функции abs()
# чтобы применить его к колонке используем метод apply()

data['days_employed'] = data['days_employed'].apply(abs) 

# проверим минимальное значение

print('Минимальный трудовой стаж:', data['days_employed'].min())

Минимальный трудовой стаж: 24.14163324048118


Приведем данные столбца days_employed (трудовой стаж) к единой временной еденице - дню.  

In [10]:
# Для этого все значения, которые больше максимального значения - 401755 часов = 16 739 дней (возьмем чуть больше - 16750) разделим на 24 часа 

data.loc[data['days_employed'] > 16750, 'days_employed'] = data.loc[data['days_employed'] > 16750, 'days_employed'] / 24

# проверим максимальное значение псосле операции
print('Максимальный трудовой стаж:', data['days_employed'].max(), 'дней')

Максимальный трудовой стаж: 16739.80835313875 дней


Для столбцов days_employed и total_income (общий доход) пропуски будут заменены релевантными значениями. В качестве среднего значения в выборке для значений, которые заместят пропуски лучше подойдет медиана - это позволит избежать искажения данных из-за выбросов. Для увеличения релевантности медиана будет рассчитана и ее значение заменит пропуски в каждом типе занятости.

In [11]:
print('Количество пропусков в столбце days_employed до обработки:', data['days_employed'].isna().sum())
print('Медианный трудовой стаж', data['days_employed'].median())

# рассчитаем медиану столбца days_employed для каждого типа занятости и сохраним в таблице med_days_employed

median_employed = data.groupby(['income_type']).agg({'days_employed': 'median'}).rename(columns = {'days_employed':'med_days_employed'})

# заменим пустные строки в столбце days_employed на медианные значения для каждого типа занятости 

data = data.merge(median_employed, on = ['income_type'])
data.loc[data['days_employed'].isna(), 'days_employed'] = data.loc[data['days_employed'].isna(), 'med_days_employed']

# проверим результат

print('Количество пропусков в столбце days_employed после обработки:', data['days_employed'].isna().sum()) 
print('Медианный трудовой стаж', data['days_employed'].median())

Количество пропусков в столбце days_employed до обработки: 2174
Медианный трудовой стаж 2194.2072712708277
Количество пропусков в столбце days_employed после обработки: 0
Медианный трудовой стаж 1991.3955312865514


<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Кстати, посмотри [тут](https://stackoverflow.com/a/53339320), как заполнить пропуски по значению другого столбца
</div>

<div class="alert alert-info">
    
Привет, Ярослав! Спасибо, учту.
</div>

<div class="alert alert-success">
    
<b> Комментарий от ревьюера 2 ✔️</b>

👍

</div>


In [12]:
# посмотрим основные статистические показатели по столбцу total_income
print('Ежемесячный доход клиентов:')
print('Максимальный доход', data['total_income'].max())
print('Минимальный доход', data['total_income'].min())
print('Средний доход', data['total_income'].mean())
print('Медианный доход', data['total_income'].median())

Ежемесячный доход клиентов:
Максимальный доход 2265604.028722744
Минимальный доход 20667.26379327158
Средний доход 167422.30220817297
Медианный доход 145017.93753253992


Заменим пропуски в столбце total_income на медианные значения ежемесячного дохода.

In [13]:
print('Количество пропусков в столбце total_income до обработки:', data['total_income'].isna().sum())
print('Медианный ежемесячный доход', data['total_income'].median())

# рассчитаем медиану столбца total_income для каждого типа занятости 
# для этого создадим датафрейм со значениями медианы для каждого типа занятости 

median_income = (data.groupby(['income_type']).agg({'total_income': 'median'}).rename(columns = {'total_income':'med_total_income'}))

# заменим пустные строки в столбце total_income на медианные значения для каждого типа занятости 
# для этого объединим датафрейм data с даатфреймом median_income с помощью метода merge()

data = data.merge(median_income, on = ['income_type'])
data.loc[data['total_income'].isna(), 'total_income'] = data.loc[data['total_income'].isna(), 'med_total_income']

print('Количество пропусков в столбце total_income после обработки:', data['total_income'].isna().sum()) 
print('Медианный ежемесячный доход', data['total_income'].median())

Количество пропусков в столбце total_income до обработки: 2174
Медианный ежемесячный доход 145017.93753253992
Количество пропусков в столбце total_income после обработки: 0
Медианный ежемесячный доход 142594.39684740017


В датафрейме появилися дополнительные столбцы с медианными значениями med_days_employed и med_total_income

In [14]:
data.head(3)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   

     family_status  family_status_id gender income_type  debt   total_income  \
0  женат / замужем                 0      F   сотрудник     0  253875.639453   
1  женат / замужем                 0      F   сотрудник     0  112080.014102   
2  женат / замужем                 0      M   сотрудник     0  145885.952297   

                   purpose  med_days_employed  med_total_income  
0            покупка жилья        1573.956856     142594.396847  
1  приобретение автомобиля        1573.956856     142594.396847  
2            покупка жилья        1573.956856     142594.396847

Минимальное значение в столбце dob_years (возраст клиента в годах) равняется 0, это значит, что часть значений были заполнены нулями вместо NaN.  

In [15]:
# посмотрим основные статистические показатели по столбцу dob_years
print('Возраст клиентов в годах:')
print('Самый страший:', data['dob_years'].max())
print('Самый младший:', data['dob_years'].min())
print('Средний возраст:', data['dob_years'].mean())
print('Медианный возраст:', data['dob_years'].median())
print('Количество значений с 0:', data.loc[data['dob_years'] == 0, 'dob_years'].count())

Возраст клиентов в годах:
Самый страший: 75
Самый младший: 0
Средний возраст: 43.29337979094077
Медианный возраст: 42.0
Количество значений с 0: 101


Значения с нулем занимают меньше 0,5% от всех значений, поэтому безболезненно заменим нули на медианное значение столбца dob_years (медианный возраст).

In [16]:
data.loc[data['dob_years'] == 0, 'dob_years'] = data['dob_years'].median()

# посмотрим основные статистические показатели по столбцу dob_years после замены
print('Возраст клиентов в годах:')
print('Самый страший:', data['dob_years'].max())
print('Самый младший:', data['dob_years'].min())
print('Средний возраст:', data['dob_years'].mean())
print('Медианный возраст:', data['dob_years'].median())
print('Количество значений с 0 после замены на медианное значение столбца:', data.loc[data['dob_years'] == 0, 'dob_years'].count())

Возраст клиентов в годах:
Самый страший: 75.0
Самый младший: 19.0
Средний возраст: 43.490452961672474
Медианный возраст: 42.0
Количество значений с 0 после замены на медианное значение столбца: 0


Посмотрим значения в столбце gender (пол клиента).

In [17]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Значение XNA в столбце gender может быть плейсхолдером пропуска значения. Поскольку предположить пол человека не представляется возможным, данные по полу клиентов не будут использованы в дальнейшем анализе, а само значение в единственном экземпляре, целесообраным считаю ничего с ним не предпринимать.  

**Выводы**

1. В стобце days_employed (общий трудовой стаж в днях) отрицательный знак значений был заменем на положительный. Все значения были приведены к единой временной мере исчисления - дню
2. В стобцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход) пропуски были заменены на медианные значения.    
3. В столбце dob_years (возраст клиента в годах) нулевые значения были заменены на медианные значения.  
4. В столбце gender был оставлен без изменений один пропуск.

### Замена типа данных

После проведенной обработки, проверим общую информацию о датасете.

In [18]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 14 columns):
children             21525 non-null int64
days_employed        21525 non-null float64
dob_years            21525 non-null float64
education            21525 non-null object
education_id         21525 non-null int64
family_status        21525 non-null object
family_status_id     21525 non-null int64
gender               21525 non-null object
income_type          21525 non-null object
debt                 21525 non-null int64
total_income         21525 non-null float64
purpose              21525 non-null object
med_days_employed    21525 non-null float64
med_total_income     21525 non-null float64
dtypes: float64(5), int64(4), object(5)
memory usage: 2.5+ MB


Для корректности дальнейших расчетов и удобства их восприятия заменим в столбцах days_employed и total_income вещественный тип данных на целочисленный.

In [19]:
# для этого используем метод astype(), т.к. мы изменяем числовые данные на числовые, 
# а не строковые на числовые, для чего применяют метод to_numeric()

data['total_income'] = data['total_income'].astype('int')
data['days_employed'] = data['days_employed'].astype('int')

data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 14 columns):
children             21525 non-null int64
days_employed        21525 non-null int64
dob_years            21525 non-null float64
education            21525 non-null object
education_id         21525 non-null int64
family_status        21525 non-null object
family_status_id     21525 non-null int64
gender               21525 non-null object
income_type          21525 non-null object
debt                 21525 non-null int64
total_income         21525 non-null int64
purpose              21525 non-null object
med_days_employed    21525 non-null float64
med_total_income     21525 non-null float64
dtypes: float64(3), int64(6), object(5)
memory usage: 2.5+ MB


**Вывод**

В столбце total_income был изменен тип данных с вещественного на целочисленный.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Здорово, что выбрал оптимальные типы данных для признаков
</div>

### Обработка дубликатов

Проверим данные на явные дубликаты и удалим их, при наличии.

In [20]:
# при помощи функции diplicated() и подсчитаем их количество функцией sum()

print('Явных дубликатов до:', data.duplicated().sum())

# удалим дуликаты. Для сохранения порядка индекццсации применим функцию reset_index()
data = data.drop_duplicates().reset_index(drop=True)

print('Явных дубликатов после:', data.duplicated().sum())

Явных дубликатов до: 55
Явных дубликатов после: 0


Проверим данные на неявные дубликаты.  
Начнем со столбца children (количество детей в семье).

In [21]:
# для этого проведем группировку по типам значений с подсчетом суммы значений для каждого

print('Количество детей в семье до обработки значений:')
data['children'].value_counts()

Количество детей в семье до обработки значений:


 0     14106
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Количество детей в семье содержит в себе значения от 0 до 5. Значение -1 и 20 являются аномалиями.   
Заменим отрицательные - на + для значения 1. 
Затем значение 20 на 2.

In [22]:
# для этого вернем абсолютные значения - 1 

data['children'] = data['children'].apply(abs) 

# используем замену с помощью метода replace().

data['children'] = data['children'].replace(20, 2)

# проверим результат

print('Количество детей в семье после обработки значений:')
data['children'].value_counts()

Количество детей в семье после обработки значений:


0    14106
1     4856
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

Посмотрим уникальные значения в столбце education (уровень образования клиента) и family_status (семейное положение). Эти столбцы, как и столбеце children, содержат типические категории, которые должны иметь одинаковый вид. Столбец purpose содержит уникальные для каждой строки значение, соответствующее индивидуальной цели клиента и может дублироваться только вместе со всей строкой. С его содержанием мы будем работать дальше с помощью лемматизации.

In [23]:
print('Столбец education:')
print(data['education'].value_counts())
print()
print('Столбец family_status:')
print(data['family_status'].value_counts())
print()

Столбец education:
среднее                13704
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Столбец family_status:
женат / замужем          12344
гражданский брак          4162
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64



Уникальных категорий, описывающих уровень образования клиента, в столбце education 5 видов: начальное, среднее, неконченное высшее, высшее и ученая степень. Остальные являются дубликатами.  

В столбце family_status дубликатов нет, но в категории "Не женат / не замужем" использован верхний регистр, что является нарушением стиля.  

В случае столбцов education и family_status проблему можно решить приведением строк к нижнему регистру. 

In [24]:
# приведем строки к одному регистру

data['education'] = data['education'].str.lower()
data['family_status'] = data['family_status'].str.lower()

# проверим, что все значения в нижнем регистре
print("Столбец education:")
print(data['education'].value_counts())
print()
print("Столбец family_status:")
print(data['family_status'].value_counts())

Столбец education:
среднее                15187
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Столбец family_status:
женат / замужем          12344
гражданский брак          4162
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64


После изменения регистра в столбце education могли появиться новые явные дубликаты. Проверим это и удалим их при наличии.

In [25]:
print('Явных дубликатов после изменения регистра:', data.duplicated().sum())

Явных дубликатов после изменения регистра: 17


In [26]:
# удалим дубликаты
data = data.drop_duplicates().reset_index(drop = True)
print('Количество оставшихся явных дубликатов:',data.duplicated().sum())

Количество оставшихся явных дубликатов: 0


**Выводы**  
  
1. Было удалено 72 явных дубликата.
2. В столбце children (количество детей в семье) значения -1 были замененены на 1, значения 20 на 2.  
3. В столбцах education (уровень образования клиента) и family_status (семейное положение) был исправлен стиль значений, приведением к нижнему регистру.
4. Дубликаты в столбце purpose необходимо убрать с помощью лемматизации.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
С обработкой явных и неявных дубликатов все отлично
</div>

### Лемматизация

Посмотрим уникальные значения в столбце purpose (цель получения кредита)

In [27]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'образование',
       'на проведение свадьбы', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку своего автомобиля',
       'на покупку подержанного автомобиля', 'сыграть свадьбу', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'получение образования', 'заняться образованием', 'автомобиль',
       'свадьба', 'получение дополнительного образования',
       'покупка недвижимости', 'операции с жильем',
       'сделка с подержанным автомобилем', 'покупка своего жилья',
       'операции с недвижимостью', 'свой автомобиль',
       'профильное образование', 'высшее образование',
       'сделка с автомобилем', 'покупка жилой недвижимости',
       'на покупку автомобиля', 'покупка жилья для сдачи',
       'операции с коммерческой недвижимостью',
     

В списке уникальных значений визуально можно выделить пять основных категории целей: недвижимость, автомобиль, образование, свадьба, ремонт. Каждая категория целей имеет несколько дубликатов. Проверим эти категории, используя метод лемматизации, т.к. в сравнени с другим методом - стеммингом - это более продвинутый алгоритм для поиска слов, записанных в разных формах.

In [28]:
# импортируем модуль Mystem из библиотеки pymystem3

from pymystem3 import Mystem

# сохраняем функцию Mystem() в переменной m

m = Mystem() 

# склеим все значения столбца purpose в строку

purpose_lemmas = ' '.join(data['purpose'])

# лемматизируем строку в перменной purpose_lemmas и сохраним результат (список лемм) в переменную lemmas

lemmas = m.lemmatize(purpose_lemmas)

# импортируем контейнер Counter из модуля collections

from collections import Counter

# подсчитаем количество лемм 

Counter(lemmas)


Counter({'покупка': 5897,
         ' ': 55020,
         'жилье': 4460,
         'приобретение': 461,
         'автомобиль': 4306,
         'дополнительный': 906,
         'образование': 4013,
         'на': 2221,
         'проведение': 767,
         'свадьба': 2323,
         'для': 1289,
         'семья': 638,
         'коммерческий': 1311,
         'недвижимость': 6351,
         'строительство': 1878,
         'собственный': 635,
         'свой': 2230,
         'подержать': 846,
         'сыграть': 765,
         'операция': 2604,
         'со': 627,
         'получение': 1314,
         'заниматься': 904,
         'с': 2918,
         'сделка': 941,
         'профильный': 436,
         'высокий': 1374,
         'жилой': 1230,
         'сдача': 651,
         'ремонт': 607,
         'подержанный': 118,
         '\n': 1})

<div class="alert alert-warning">
    
<b> Комментарий от ревьюера ⚠️</b>
    
Все библиотеки лучше импортировать в начале файла. Об этом говорится в [стилевом гайде по питону](https://www.python.org/dev/peps/pep-0008/#imports)
    
    Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.
</div>

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Можно использовать метод most_common, чтобы упорядочить значения Counter. Так будет проще выделить основные категории
    
`Counter(lemmas).most_common()`
</div>

<div class="alert alert-info">
    
Спасибо, учту.
</div>

Подсчет лемм подтвердил визуально выделенные категории.  

**Вывод**

1. С помощью лемматизации выделены 5 категорий целей взятия кредита:  
 1) Недвижимость  
 2) Автомобиль  
 3) Образование  
 4) Свадьба  
 5) Ремонт  

### Категоризация данных

**Категоризация по целям взятия кредита**

Создадим новый столбец purpose_category с категориями, соответствующими каждой цели в столбце purpose.

In [29]:
# для этого создадим функцию purpose_category

'''
Функция purpose_category возвращает название категории для каждой цели.
В качестве аргумента принимает строку.
'''

def purpose_category(row):
    lemmas = m.lemmatize(row['purpose'])
    for row in lemmas:
        if 'ремонт' in lemmas:
            return 'ремонт'
        if ('недвижимость' in lemmas or 
            'жилье' in lemmas):
            return 'недвижимость'
        if 'автомобиль' in lemmas:
            return 'автомобиль'
        if 'образование' in lemmas:
            return 'образование'
        if 'свадьба' in lemmas:
            return 'свадьба'
    return 'другая категория'


# для применения функции purpose_category к столбцу используем функцию apply() c аргументом axis=1 (применить к строкам)

data['purpose_category'] = data.apply(purpose_category, axis=1)

# подсчитаем значения в столбце
print('Количество категорий в столбце purpose_category:')
print(data['purpose_category'].value_counts())
data.head(3)

Количество категорий в столбце purpose_category:
недвижимость    10204
автомобиль       4306
образование      4013
свадьба          2323
ремонт            607
Name: purpose_category, dtype: int64


children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   

     family_status  family_status_id gender income_type  debt  total_income  \
0  женат / замужем                 0      F   сотрудник     0        253875   
1  женат / замужем                 0      F   сотрудник     0        112080   
2  женат / замужем                 0      M   сотрудник     0        145885   

                   purpose  med_days_employed  med_total_income  \
0            покупка жилья        1573.956856     142594.396847   
1  приобретение автомобиля        1573.956856     142594.396847   
2            покупка жилья        1573.956856     142594.396847   

  purpose_category  
0     недвижимость  
1       автомобиль  
2     недвижимость

**Категоризация по ежемесячному доходу**  

Для начала посмотрим статистические данные по столбцу total_income (ежемесячный доход).

In [30]:
# используем метод format(), чтобы изменить количество знаков после точки для удобства восприятия данных

data['total_income'].describe().apply("{:.2f}".format)

count      21453.00
mean      165320.63
std        98189.47
min        20667.00
25%       107620.00
50%       142594.00
75%       195821.00
max      2265604.00
Name: total_income, dtype: object

Итак, средний доход: 165.320, среднее отклонение значений в нашей выборке 98.189.  
Исходя из этих статистических показателей (округлив их значения), категоризируем уровень дохода на 4 категории по уровню дохода:

доход меньше 67.000 рублей — категория «доход ниже среднего»;  
доход от 67.000 до 143.000 рублей — категория «средний доход»;  
доход от 143.000 до 263.000 — категория «доход выше среднего»;  
доход выше 263.000 рублей — категория «высокий доход».  
  
Создадим столбец income_category с названиями данных категорий, каждая из которых соответствует доходу в столбце total_income.

In [31]:
# для этого создадим функцию income_category

'''
Функция income_category возвращает название категории для ежемесячного дохода в стобце total_income.
В качестве аргумента принимает строку из столбца total_income.
'''

def income_category(row):
        income = row['total_income']
        if income <= 67000:
            return 'доход ниже среднего'
        if income > 67000 and income <= 143000:
            return 'средний доход'
        if income > 143001 and income <= 263000:
            return 'доход выше среднего'
        if income > 263001:
            return 'высокий доход'

data['income_category'] = data.apply(income_category, axis = 1)
print(data['income_category'].value_counts())
data.head(3)

средний доход          9615
доход выше среднего    8222
высокий доход          2344
доход ниже среднего    1272
Name: income_category, dtype: int64


children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   

     family_status  family_status_id gender income_type  debt  total_income  \
0  женат / замужем                 0      F   сотрудник     0        253875   
1  женат / замужем                 0      F   сотрудник     0        112080   
2  женат / замужем                 0      M   сотрудник     0        145885   

                   purpose  med_days_employed  med_total_income  \
0            покупка жилья        1573.956856     142594.396847   
1  приобретение автомобиля        1573.956856     142594.396847   
2            покупка жилья        1573.956856     142594.396847   

  purpose_category      income_category  
0     недвижимость  доход выше среднего  
1       автомобиль        средний доход  
2     недвижимость  доход выше среднего

**Категоризация по количеству детей**  

Категоризируем клиентов по количеству детей:  
- "без детей";  
- "1-2 ребенка";  
- "более 3 детей"  
  
Создаадим новый столбец children_category с данными категориями.

In [32]:
# для этого создадим функцию children_category

'''
Функция income_category возвращает название категории для ежемесячного дохода в стобце total_income.
В качестве аргумента принимает строку из столбца total_income.
'''

def children_category(row):
        child = row['children']
        if child == 0:
            return 'без детей'
        if child >= 1 and child <= 2:
            return '1-2 ребенка'
        if child >= 3:
            return 'более 3 детей'

data['children_category'] = data.apply(children_category, axis = 1)
print(data['children_category'].value_counts())
data.head(3)

без детей        14090
1-2 ребенка       6983
более 3 детей      380
Name: children_category, dtype: int64


children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   

     family_status  family_status_id gender income_type  debt  total_income  \
0  женат / замужем                 0      F   сотрудник     0        253875   
1  женат / замужем                 0      F   сотрудник     0        112080   
2  женат / замужем                 0      M   сотрудник     0        145885   

                   purpose  med_days_employed  med_total_income  \
0            покупка жилья        1573.956856     142594.396847   
1  приобретение автомобиля        1573.956856     142594.396847   
2            покупка жилья        1573.956856     142594.396847   

  purpose_category      income_category children_category  
0     недвижимость  доход выше среднего       1-2 ребенка  
1       автомобиль        средний доход       1-2 ребенка  
2     недвижимость  доход выше среднего         без детей

**Выводы**

1. Цели взятия кредита были категоризированы в 5 категорий. Для них был создан столбец purpose_category.  
2. Ежемесячный доход категоризирован в 4 категории. Для них был создан столбец income_category.  
3. Количество детей в семье категоризированы в 3 категории. Для них был создан столбец children_category.  
  
Предобработка данных завершена, данные подготовлены для исследовательского анализа.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
С добавлением новых признаков все отлично!
</div>


## Шаг 3. Ответьте на вопросы

Значения названий колонок в сводных таблицах:  
  
**...category** - названия категорий;  
**count debt**  -  количество заемщиков;  
**sum debt**  -  количество имевших задолженность;  
**ratio_debt** - доля имевших задолженность в процентах.
		
		

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [33]:
# Для сравнения наблюдений будем использовать метод свободных таблиц pivot_table().
# Аргументы метода:
    # index — столбец или столбцы, по которым группируют данные 
    # values — значения, по которым мы хотим увидеть сводную таблицу 
    # aggfunc — функция, применяемая к значениям
    
# Создадим таблицу children_pivot.    
children_pivot = data.pivot_table(index = 'children_category', 
                                          values = 'debt', # количество задолженностей по категории
                                          aggfunc = ['count','sum']) 

# Создадим в таблице children_pivot новый столбец 'ratio_debt'. 
# Cохраним в нём процент невернувших кредит от всех заемщиков каждой категории.
# Чем больше значение в столбце 'ratio_debt', тем более неблагонадежна категория заемщиков. 

# Для лучшего представления данных в столбце ratio_debt у множим результат деления на 100, округлим с помощью функции round()
# Далее используем функцию astype() для форматирования типа данных float в строковый тип 

children_pivot['ratio_debt'] = (children_pivot['sum'] / children_pivot['count'] * 100).round(1).astype('str') +'%'

# Выведем таблицу

children_pivot

count   sum ratio_debt
                    debt  debt           
children_category                        
1-2 ребенка         6983   647       9.3%
без детей          14090  1063       7.5%
более 3 детей        380    31       8.2%

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Если мы делим сумму на количество, то получаем среднее значение. Его можно посчитать при группировке данных. Знак % можно добавить с помощью форматирования признаков
    
```
(
    data
    .pivot_table(index = 'children_category', 
                 values = 'debt', # количество задолженностей по категории
                 aggfunc = ['count','sum','mean'])
    .style.format({('mean','debt') : '{:.2%}'})
)
```
</div>

<div class="alert alert-info">
    
Спасибо, учту.
</div>

Посмотрим долю невозвратов не по категориям, а по каждому значению количества детей от 0 до 5.

In [34]:
children_pivot_2 = data.pivot_table(index = 'children', 
                                          values = 'debt', 
                                          aggfunc = ['count','sum']) 

children_pivot_2['ratio_debt'] = (children_pivot_2['sum'] / children_pivot_2['count'] * 100).round(1).astype('str') +'%'
children_pivot_2

count   sum ratio_debt
           debt  debt           
children                        
0         14090  1063       7.5%
1          4855   445       9.2%
2          2128   202       9.5%
3           330    27       8.2%
4            41     4       9.8%
5             9     0       0.0%

**Вывод**

Имеется зависимость между наличием детей и возвратом кредита в срок. 

Лучшие заемщики - клиенты без детей (7,5% имели задолженность). Также эта категория составляет 65% от всех заемщиков, что делает этот процент существенно более весомым.   

Средние по надежности заемщики - это клиенты с количеством детей 3 и больше (8,2% имели задолженность).  

Самые ненадежные заемщики - клиенты со средним количеством детей (1-2) (9,3% имели задолженность).  

В более детальном анализе самый высокий процент имевших задолженность - 9,8% - имеют заемщики с 4 детьми, но их выборка недостаточна (41 человек) для полноценного сравнения с другими категориями.  

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Ты хорошо справился с определением зависимости задолженности от наличия детей
</div>


### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [35]:
# Создадим таблицу family_status_pivot.    
family_status_pivot = data.pivot_table(index = 'family_status', 
                                          values = 'debt', # количество задолженностей по категории
                                          aggfunc = ['count','sum']) 

# Создадим в таблице family_status_pivot новый столбец 'ratio_debt'. 

family_status_pivot['ratio_debt'] = (family_status_pivot['sum'] / family_status_pivot['count'] * 100).round(1).astype('str') +'%'

# Выведем таблицу

family_status_pivot

count  sum ratio_debt
                        debt debt           
family_status                               
в разводе               1195   85       7.1%
вдовец / вдова           959   63       6.6%
гражданский брак        4150  388       9.3%
женат / замужем        12339  931       7.5%
не женат / не замужем   2810  274       9.8%

**Вывод**

Имеется зависимость между семейным положением и возвратом кредита в срок.  
  
Лучшие заемщики - вдовы / вдовцы (6,6% имели задолженность).  
К этой же категории клиентов - самых надежных заемщиков - можно отнести также заемщиков, состоящих в официальном браке. Не смотря на то, что процент имевших задолженность в этой категории выше - 7,5%, надо иметь ввиду, что эта группа клиентов составляет 57% ото всех заемщиков.  Вполне вероятно, что, если бы их количество было бы меньше, то и процент имевших задолженность уменьшился бы.  
  
Средние по надежности заемщики - это заемщики в разводе (7,1% имели задолженность).  

Самые ненадежные заемщики представлены 2 категориями клиентов: заемщики в гражданском браке (9,3% имели задолженность) и клиенты, несостоящие в официальном браке (9,8% имели задолженность).  

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [36]:
# Создадим таблицу family_status_pivot.    
income_pivot = data.pivot_table(index = 'income_category', 
                                          values = 'debt', # количество задолженностей по категории
                                          aggfunc = ['count','sum']) 

# Создадим в таблице family_status_pivot новый столбец 'ratio_debt'. 

income_pivot['ratio_debt'] = (income_pivot['sum'] / income_pivot['count'] * 100).round(1).astype('str') +'%'

# Выведем таблицу

income_pivot

count  sum ratio_debt
                     debt debt           
income_category                          
высокий доход        2344  163       7.0%
доход выше среднего  8222  662       8.1%
доход ниже среднего  1272   90       7.1%
средний доход        9615  826       8.6%

**Вывод**

Имеется зависимость между уровнем дохода и возвратом кредита в срок.  

Лучшие заемщики представлены 2 категориями: клиенты с доходом выше среднего (больше 263.000 рублей/мес.) (7% имели задолженность) и с доходом ниже среднего (доход меньше 67.000 рублей/мес.) (7,1% имели задолженность). 

Ненадежные заемщики также представлены 2 категориями: клиенты со средним уровнем дохода (доход от 67.000 до 143.000 рублей/мес.) (8,6% имели задолженность) и клиенты с уровнем дохода выше среднего (доход от 143.000 до 263.000 рублей/мес.) (8,1% имели задолженность).

### Как разные цели кредита влияют на его возврат в срок?

In [37]:
# Создадим таблицу family_status_pivot.    
purpose_pivot = data.pivot_table(index = 'purpose_category', 
                                          values = 'debt', # количество задолженностей по категории
                                          aggfunc = ['count','sum']) 

# Создадим в таблице family_status_pivot новый столбец 'ratio_debt'. 

purpose_pivot['ratio_debt'] = (purpose_pivot['sum'] / purpose_pivot['count'] * 100).round(1).astype('str') +'%'

# Выведем таблицу

purpose_pivot

count  sum ratio_debt
                   debt debt           
purpose_category                       
автомобиль         4306  403       9.4%
недвижимость      10204  747       7.3%
образование        4013  370       9.2%
ремонт              607   35       5.8%
свадьба            2323  186       8.0%

**Вывод**

Самые надежные заемщики те, кто в качестве целей взятия кредита указали цели из категории "Ремонт" (5,8% имели задолженность). Правда, здесь нужно учитывать существенно меньшее количество клиентов в этой категории целей по сравнению с другими (2,8% от всей выборки).  

Учитывая малый размер первой группы заемщиков, мы бы предположили, что не менее надежными являются клиенты с целями взятия кредита из категории "Недвижимость" (7,3% имели задолженность). С целями из этой категории, напротив, самая многочисленная группа заемщиков - 47% всей выборки.  

Средние по надежности заемщики - это заемщики с целями из категории "Свадьба" (8% имели задолженность).  

Самые ненадежные заемщики представлены 2 категориями клиентов: заемщики с целями из категории "Образование" (9,2% имели задолженность) и клиенты с целями из категории "Автомобиль" (9,4% имели задолженность).  

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Все верно
</div>

## Шаг 4. Общий вывод

Данный анлиз выявил наличие зависимости возврата кредита в срок от следующих характеристик заемщика:  
- количества детей в семье;  
- семейного положения;   
- уровня дохода;  
- цели взятия кредита.  
  
Идеальный заемщик по этим факторам - это клиент в разводе, либо вдовец / вдова, либо состоящий в официальном браке, без детей, с доходом ниже среднего или высоким, который берет кредит либо на свадьбу, либо на операции с недвижимостью.  
  
Самый же ненадежный средний заемщик - это клиент, несостоящий в официальной браке, либо состоящий в гражданском браке, имеющий 1-2 детей, доход средний или выше среднего, который берет кредит на образование или покупку автомобиля.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Хороший вывод
</div>



<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Спасибо за твою работу! Ты хорошо справился с подготовкой и анализом данных.

Критических замечаний нет, но я верну тебе работу на случай, если у тебя появились вопросы. Если вопросов нет, можешь прислать работу без изменений, и я ее приму.
</div>

<div class="alert alert-info">
    
Ярослав, спасибо за содержательное ревью! Вопросов нет.
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2 ✔️</b>

Надеюсь, эта работа была полезна для тебя. Я принимаю проект и желаю удачи в прохождении курса!

</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.


- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.